File: src/main/java/com/DavidM1A2/afraidofthedark/common/spell/component/powerSource/SpellPowerSourceExperience.java
package com.DavidM1A2.afraidofthedark.common.spell.component.powerSource;

import com.DavidM1A2.afraidofthedark.common.constants.ModSpellPowerSources;
import com.DavidM1A2.afraidofthedark.common.spell.Spell;
import com.DavidM1A2.afraidofthedark.common.spell.component.powerSource.base.AOTDSpellPowerSource;
import com.DavidM1A2.afraidofthedark.common.spell.component.powerSource.base.SpellPowerSourceEntry;
import net.minecraft.entity.player.EntityPlayer;

/**
 * Class representing the experience source
 */
public class SpellPowerSourceExperience extends AOTDSpellPowerSource
{
    // The number of units each xp level supplies
    private static final double UNIT_COST_PER_LEVEL = 5;

    /**
     * Constructor just calls super
     */
    public SpellPowerSourceExperience()
    {
        super();
    }

    /**
     * True if the given spell can be cast, false otherwise
     *
     * @param entityPlayer The player that is casting the spell
     * @param spell        The spell to attempt to cast
     * @return True if the spell can be cast, false otherwise
     */
    @Override
    public boolean canCast(EntityPlayer entityPlayer, Spell spell)
    {
        int xpLevelCost = (int) Math.ceil(spell.getCost() / UNIT_COST_PER_LEVEL);
        return xpLevelCost <= entityPlayer.experienceLevel;
    }

    /**
     * Does nothing, creative power sources don't use energy
     *
     * @param entityPlayer The player that is casting the spell
     * @param spell        the spell to attempt to cast
     */
    public void consumePowerToCast(EntityPlayer entityPlayer, Spell spell)
    {
        int xpLevelCost = (int) Math.ceil(spell.getCost() / UNIT_COST_PER_LEVEL);
        entityPlayer.addExperienceLevel(-xpLevelCost);
    }

    /**
     * Gets a description message of how cost is computed for this power source
     *
     * @return A description describing how cost is computed
     */
    @Override
    public String getCostDescription()
    {
        return "1xp level for every " + UNIT_COST_PER_LEVEL + " units of spell cost";
    }

    /**
     * Computes the message describing why the power source doesn't have enough power
     *
     * @return A string describing why the power source doesn't have enough energy
     */
    @Override
    public String getUnlocalizedOutOfPowerMsg()
    {
        return "aotd.spell.power_source.experience.invalid_msg";
    }

    /**
     * Should get the SpellPowerSourceEntry registry's type
     *
     * @return The registry entry that this power source was built with, used for deserialization
     */
    @Override
    public SpellPowerSourceEntry getEntryRegistryType()
    {
        return ModSpellPowerSources.EXPERIENCE;
    }
}

File: src/main/kotlin/com/davidm1a2/afraidofthedark/common/spell/component/powerSource/SpellPowerSourceExperience.kt
package com.davidm1a2.afraidofthedark.common.spell.component.powerSource

import com.davidm1a2.afraidofthedark.common.constants.Constants
import com.davidm1a2.afraidofthedark.common.constants.LocalizationConstants
import com.davidm1a2.afraidofthedark.common.spell.Spell
import com.davidm1a2.afraidofthedark.common.spell.component.powerSource.base.AOTDSpellPowerSource
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.ResourceLocation
import kotlin.math.ceil

/**
 * Class representing the experience source
 */
class SpellPowerSourceExperience : AOTDSpellPowerSource(ResourceLocation(Constants.MOD_ID, "experience")) {
    /**
     * True if the given spell can be cast, false otherwise
     *
     * @param entity The entity that is casting the spell
     * @param spell        The spell to attempt to cast
     * @return True if the spell can be cast, false otherwise
     */
    override fun canCast(entity: Entity, spell: Spell): Boolean {
        return (entity as? EntityPlayer)?.let {
            val xpLevelCost = ceil(spell.getCost() / UNIT_COST_PER_LEVEL).toInt()
            return xpLevelCost <= it.experienceLevel
        } ?: false
    }

    /**
     * Does nothing, creative power sources don't use energy
     *
     * @param entity The entity that is casting the spell
     * @param spell        the spell to attempt to cast
     */
    override fun consumePowerToCast(entity: Entity, spell: Spell) {
        (entity as? EntityPlayer)?.let {
            val xpLevelCost = ceil(spell.getCost() / UNIT_COST_PER_LEVEL).toInt()
            it.addExperienceLevel(-xpLevelCost)
        }
    }

    /**
     * Gets a description message of how cost is computed for this power source
     *
     * @return A description describing how cost is computed
     */
    override fun getCostDescription(): String {
        return "1xp level for every $UNIT_COST_PER_LEVEL units of spell cost"
    }

    /**
     * Computes the message describing why the power source doesn't have enough power
     *
     * @return A string describing why the power source doesn't have enough energy
     */
    override fun getUnlocalizedOutOfPowerMsg(): String {
        return LocalizationConstants.Spell.POWER_SOURCE_EXPERIENCE_INVALID_MSG
    }

    companion object {
        // The number of units each xp level supplies
        private const val UNIT_COST_PER_LEVEL = 5.0
    }
}